File: internal/memory/memory.go
// Package memory 智能记忆：会话后事实抽取、向量去重、批量写入、画像聚合、衰减、隐私模式。
package memory

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"

	"github.com/helixnow/deep-student-go/pkg/llm"
	"github.com/helixnow/deep-student-go/pkg/store"
	"github.com/helixnow/deep-student-go/pkg/vfs"
)

// Action 记忆操作决策。
type Action string

const (
	ActionAdd    Action = "ADD"
	ActionUpdate Action = "UPDATE"
	ActionAppend Action = "APPEND"
	ActionDelete Action = "DELETE"
	ActionNone   Action = "NONE"
)

// Item 单条记忆。
type Item struct {
	ID        string            `json:"id"`
	Category  string            `json:"category"` // identity | preference | goal | subject | other
	Content   string            `json:"content"`
	Tags      []string          `json:"tags"`
	Weight    int               `json:"weight"`
	HitCount  int               `json:"hit_count"`
	LastHit   time.Time         `json:"last_hit"`
	CreatedAt time.Time         `json:"created_at"`
	UpdatedAt time.Time         `json:"updated_at"`
	Source    string            `json:"source,omitempty"`
	Metadata  map[string]string `json:"metadata,omitempty"`
}

// Service 记忆服务。
type Service struct {
	vfs   *vfs.FS
	store *store.Store
	llm   *llm.Registry
	mu    sync.Mutex
	items map[string]*Item
}

// New 创建 Service。
func New(fs *vfs.FS, st *store.Store, l *llm.Registry) *Service {
	return &Service{vfs: fs, store: st, llm: l, items: map[string]*Item{}}
}

// ExtractFromConversation 从对话中抽取事实。
func (s *Service) ExtractFromConversation(ctx context.Context, conversation string) ([]string, error) {
	p, ok := s.llm.Get("openai")
	if !ok {
		return nil, fmt.Errorf("memory: no provider")
	}
	resp, err := p.Chat(ctx, llm.ChatRequest{
		Model: "gpt-4o-mini",
		Messages: []llm.Message{
			{Role: llm.RoleSystem, Content: "Extract durable facts about the user from the conversation. Categories: identity, preference, goal, subject status, other. Return a JSON array of strings, one fact per element. Skip transient or unclear content."},
			{Role: llm.RoleUser, Content: conversation},
		},
	})
	if err != nil {
		return nil, err
	}
	start := strings.Index(resp.Content, "[")
	end := strings.LastIndex(resp.Content, "]")
	var facts []string
	if start >= 0 && end > start {
		_ = json.Unmarshal([]byte(resp.Content[start:end+1]), &facts)
	}
	return facts, nil
}

// Decide 对每条事实给出 ADD/UPDATE/APPEND/DELETE/NONE 决策。
func (s *Service) Decide(ctx context.Context, fact string) (Action, string, error) {
	s.mu.Lock()
	existing := make([]*Item, 0, len(s.items))
	for _, it := range s.items {
		existing = append(existing, it)
	}
	s.mu.Unlock()
	summary := make([]string, 0, len(existing))
	for _, e := range existing {
		summary = append(summary, fmt.Sprintf("[%s] %s", e.ID, e.Content))
	}
	p, _ := s.llm.Get("openai")
	if p == nil {
		return ActionAdd, "", nil
	}
	resp, err := p.Chat(ctx, llm.ChatRequest{
		Model: "gpt-4o-mini",
		Messages: []llm.Message{
			{Role: llm.RoleSystem, Content: `Decide whether the new fact should be ADD, UPDATE, APPEND, DELETE, or NONE. Return JSON: {"action":"ADD|UPDATE|APPEND|DELETE|NONE","target_id":"<id or empty>","content":"<final content>"}`},
			{Role: llm.RoleUser, Content: "Existing memories:\n" + strings.Join(summary, "\n") + "\n\nNew fact: " + fact},
		},
	})
	if err != nil {
		return ActionNone, "", err
	}
	start := strings.Index(resp.Content, "{")
	end := strings.LastIndex(resp.Content, "}")
	var dec struct {
		Action   string `json:"action"`
		TargetID string `json:"target_id"`
		Content  string `json:"content"`
	}
	if start >= 0 && end > start {
		_ = json.Unmarshal([]byte(resp.Content[start:end+1]), &dec)
	}
	return Action(dec.Action), dec.TargetID, nil
}

// ApplyFact 应用一个事实。
func (s *Service) ApplyFact(fact string) *Item {
	now := time.Now()
	it := &Item{
		ID: uuid.NewString(), Category: "other", Content: fact,
		CreatedAt: now, UpdatedAt: now, LastHit: now, Weight: 1,
	}
	s.mu.Lock()
	s.items[it.ID] = it
	s.mu.Unlock()
	return it
}

// Ingest 抽取 + 决策 + 写入。
func (s *Service) Ingest(ctx context.Context, conversation string) ([]*Item, error) {
	facts, err := s.ExtractFromConversation(ctx, conversation)
	if err != nil {
		return nil, err
	}
	out := make([]*Item, 0, len(facts))
	for _, f := range facts {
		action, _, _ := s.Decide(ctx, f)
		switch action {
		case ActionAdd, ActionUpdate, ActionAppend:
			out = append(out, s.ApplyFact(f))
		}
	}
	return out, nil
}

// Search 简单关键词搜索。
func (s *Service) Search(q string) []*Item {
	s.mu.Lock()
	defer s.mu.Unlock()
	out := make([]*Item, 0)
	for _, it := range s.items {
		if strings.Contains(strings.ToLower(it.Content), strings.ToLower(q)) {
			it.HitCount++
			it.LastHit = time.Now()
			out = append(out, it)
		}
	}
	return out
}

// Profile 汇总为用户画像。
func (s *Service) Profile() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	byCat := map[string][]*Item{}
	for _, it := range s.items {
		byCat[it.Category] = append(byCat[it.Category], it)
	}
	var sb strings.Builder
	for cat, items := range byCat {
		sb.WriteString("## " + cat + "\n")
		for _, it := range items {
			sb.WriteString("- " + it.Content + "\n")
		}
	}
	return sb.String()
}

// Decay 90 天未用降权 / 高频加权。
func (s *Service) Decay() {
	s.mu.Lock()
	defer s.mu.Unlock()
	now := time.Now()
	for _, it := range s.items {
		days := int(now.Sub(it.LastHit).Hours() / 24)
		if days > 90 {
			it.Weight = maxInt(0, it.Weight-1)
		}
		if it.HitCount > 5 {
			it.Weight = minInt(10, it.Weight+1)
		}
	}
}

// PrivacyMode 隐私模式：阻止所有外部 API。
func (s *Service) PrivacyMode(on bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if on {
		s.llm = llm.NewRegistry() // 清空所有 provider
	}
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}
func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: frontend/src/state/memory.ts
// Memory Store —— 智能记忆状态管理
// ------------------------------------------------------------
// 对接后端 Wails 绑定：
// - MemoryIngest(conversation) → Item[]
// - MemorySearch(q) → Item[]
// - MemoryProfile() → string / MemoryPrivacyMode(on)
// - MemoryDecay()
//
// 设计要点：
// - 三栏布局：左操作面板 / 中记忆列表 / 右详情 + 摄入
// - 记忆按 category 分组：identity / preference / goal / subject / other
// - 支持搜索、用户画像、隐私模式、衰减

import { create } from "zustand";
import { callWails } from "@/lib/wails";

/** 记忆条目 —— 与后端 memory.Item 对齐 */
export interface MemoryItem {
  id: string;
  category: string; // identity | preference | goal | subject | other
  content: string;
  tags: string[];
  weight: number;
  hit_count: number;
  last_hit: string;
  created_at: string;
  updated_at: string;
  source?: string;
  metadata?: Record<string, string>;
}

/** 记忆分类 */
export const MEMORY_CATEGORIES = [
  { key: "identity", label: "身份" },
  { key: "preference", label: "偏好" },
  { key: "goal", label: "目标" },
  { key: "subject", label: "学科" },
  { key: "other", label: "其他" },
] as const;

interface MemoryState {
  /** 全部记忆（搜索结果或全量） */
  items: MemoryItem[];
  /** 搜索关键词 */
  searchQuery: string;
  /** 是否处于搜索结果模式 */
  searching: boolean;
  /** 当前选中记忆 ID */
  selectedId: string | null;
  /** 用户画像文本 */
  profile: string;
  /** 隐私模式开关 */
  privacyMode: boolean;
  /** 摄入对话文本 */
  ingestText: string;
  /** 摄入后新增的记忆 */
  ingestedItems: MemoryItem[];
  /** 加载状态 */
  loading: boolean;
  /** 画像加载状态 */
  profileLoading: boolean;
  /** 错误信息 */
  error: string | null;
  /** 操作提示 */
  toast: string | null;

  // —— Actions ——
  setSearchQuery: (q: string) => void;
  setIngestText: (t: string) => void;
  selectItem: (id: string | null) => void;
  search: () => Promise<void>;
  clearSearch: () => void;
  loadProfile: () => Promise<void>;
  ingest: () => Promise<void>;
  togglePrivacy: () => Promise<void>;
  decay: () => Promise<void>;
}

export const useMemoryStore = create<MemoryState>((set, get) => ({
  items: [],
  searchQuery: "",
  searching: false,
  selectedId: null,
  profile: "",
  privacyMode: false,
  ingestText: "",
  ingestedItems: [],
  loading: false,
  profileLoading: false,
  error: null,
  toast: null,

  setSearchQuery: (q) => set({ searchQuery: q }),
  setIngestText: (t) => set({ ingestText: t }),
  selectItem: (id) => set({ selectedId: id }),

  // —— 搜索记忆 ——
  search: async () => {
    const { searchQuery } = get();
    if (!searchQuery.trim()) {
      set({ error: "请输入搜索关键词" });
      return;
    }
    set({ loading: true, error: null, searching: true, selectedId: null });
    try {
      const list = await callWails<MemoryItem[]>("MemorySearch", searchQuery.trim());
      set({ items: list ?? [] });
    } catch (err) {
      set({ error: err instanceof Error ? err.message : String(err) });
    } finally {
      set({ loading: false });
    }
  },

  clearSearch: () =>
    set({ searching: false, items: [], searchQuery: "", selectedId: null }),

  // —— 加载用户画像 ——
  loadProfile: async () => {
    set({ profileLoading: true, error: null });
    try {
      const text = await callWails<string>("MemoryProfile");
      set({ profile: text ?? "[后端未连接] 画像不可用" });
    } catch (err) {
      set({ error: err instanceof Error ? err.message : String(err) });
    } finally {
      set({ profileLoading: false });
    }
  },

  // —— 摄入对话 ——
  ingest: async () => {
    const { ingestText } = get();
    if (!ingestText.trim()) {
      set({ error: "请粘贴对话文本" });
      return;
    }
    set({ loading: true, error: null });
    try {
      const newItems = await callWails<MemoryItem[]>(
        "MemoryIngest",
        ingestText
      );
      const items = newItems ?? [];
      set({ ingestedItems: items, ingestText: "" });
      set({ toast: `已摄入 ${items.length} 条记忆` });
      // 2 秒后清除提示
      setTimeout(() => set({ toast: null }), 2000);
    } catch (err) {
      set({ error: err instanceof Error ? err.message : String(err) });
    } finally {
      set({ loading: false });
    }
  },

  // —— 切换隐私模式 ——
  togglePrivacy: async () => {
    const { privacyMode } = get();
    const next = !privacyMode;
    set({ privacyMode: next, error: null });
    try {
      await callWails<void>("MemoryPrivacyMode", next);
      set({ toast: next ? "隐私模式已开启" : "隐私模式已关闭" });
      setTimeout(() => set({ toast: null }), 2000);
    } catch (err) {
      set({ error: err instanceof Error ? err.message : String(err) });
    }
  },

  // —— 衰减 ——
  decay: async () => {
    set({ loading: true, error: null });
    try {
      await callWails<void>("MemoryDecay");
      set({ toast: "衰减已完成" });
      setTimeout(() => set({ toast: null }), 2000);
    } catch (err) {
      set({ error: err instanceof Error ? err.message : String(err) });
    } finally {
      set({ loading: false });
    }
  },
}));
